Feature/erc777 #1159

catageek opened the pull request:
<!-- 0. 🎉 Thank you for submitting a PR! -->

<!-- 1. Does this close any open issues? Please list them below. -->

<!-- Keep in mind that new features have a better chance of being merged fast if
they were first discussed and designed with the maintainers. If there is no
corresponding issue, please consider opening one for discussion first! -->

Fixes #1159 

<!-- 2. Describe the changes introduced in this pull request. -->
<!--    Include any context necessary for understanding the PR's purpose. -->
This PR adds a [ERC777](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md) compatible token contract in the drafts folder.

A few points:

- The interfaces IERC777TokenSender.sol and IERC777TokenReceiver.sol are under MPL license since it's the license used by the reference implementation. I didn't take the responsibility to remove the license. For IERC777.sol, the license is not obvious also.
- This implementation uses a local version of test/helpers/ExpectEvent.js because the implementation provided by the openzeppelin-test-helpers package does not work for events triggered during contract creation. A PR to the repository of the helpers package is expected, and this local version will be discarded when it's merged.
- test/introspection/ERC1820Deploy.js is the script that deploys an ERC1820 registry before executing the tests.
- The test file is a big monolithic file that should be dispatched in several files. I am not a Javascript developper, so forgive me to release it "as is".
- The security of the code has not been reviewed.

<!-- 3. Before submitting, please make sure that you have:
  - reviewed the OpenZeppelin Contributor Guidelines
    (https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/CONTRIBUTING.md),
  - added tests where applicable to test new functionality,
  - made sure that your contracts are well-documented,
  - run the Solidity linter (`npm run lint:sol`) and fixed any issues,
  - run the JS linter and fixed any issues (`npm run lint:fix`), and
  - updated the changelog, if applicable.
-->

catageek commented:
OK, I drop trying to make the unit tests pass. For an unknown reason, I can't deploy an ERC1820 registry contract because it has already been deployed (nonce = 1), but I can't access the existing one and all the tests fail.

nventuro commented:
Hello @catageek! Sorry for taking so long to review this, I was out of town for a couple days.

One small comment before I review the code proper:

>The interfaces IERC777TokenSender.sol and IERC777TokenReceiver.sol are under MPL license since it's the license used by the reference implementation. I didn't take the responsibility to remove the license. For IERC777.sol, the license is not obvious also.

From my understanding of the EIP [copyright](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md#copyright), all rights are waved via CC0. This includes the contract interfaces, so we should be free to use them and re-license them under MIT (OpenZeppelin's license). IANAL though.

nventuro commented on contracts/drafts/ERC777/ERC777Base.sol:
```diff
@@ -0,0 +1,491 @@
+        uint256 amount,
+        bytes calldata data
+    ) external {
+        _send(
```
I think we should check here if `msg.sender` is allowed to do this, and have `_sender` be permissionless (same as how `ERC20`'s `transfer` and friends work).

nventuro commented:
@catageek if you don't mind, I'll start working directly on top of your commits in your branch, since there are multiple extremely minor things to polish. The overall PR looks absolutely great though, thanks a bunch! This is probably one of the best contributions to the repo I've seen: most issues come from the sheer size of it. 

I'll also look into the tests and see if I can figure out what's going on there. I'll try to keep the commits to the point so that this can be more easily reviewed.

Once again, thanks!

nventuro commented:
@catageek I'm getting permission errors when trying to push to your fork, have you disabled maintainers being able to edit your PR?

catageek commented:
Hi @nventuro. I didn't activate any restriction on my repo and I don't get why you are not authorized, anyway I added you as a collaborator. I hope it will work.
Feel free to work on my branch, I spent a lot of time on it and I moved to another personal project

nventuro commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+     * @param userData bytes extra information provided by the token holder (if any)
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function _send(
```
I think we're missing this condition here, from the spec:

```
The token contract MUST revert when minting in any of the following cases:
    The recipient is a contract, and it does not implement the ERC777TokensRecipient interface via ERC1820.
```

catageek commented on contracts/drafts/ERC777/IERC777TokensRecipient.sol:
```diff
@@ -0,0 +1,15 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+pragma solidity ^0.5.2;
+
+interface IERC777TokensRecipient {
```
I agree

catageek commented on contracts/drafts/ERC777/ERC777Base.sol:
```diff
@@ -0,0 +1,491 @@
+
+    uint256 private _granularity;
+
+    address[] private _defaultOpsArray;
```
OK

nventuro commented:
Being a collaborator worked, thanks! I'm currently rewriting some of the tests based on your code to get the more in-line with the rest of the test suite.

nventuro commented:
@frangio I'm leaning towards making the base `ERC777` have an `internal` constructor, since it makes no sense to instantiate one as-is (since there would be no token holders). The same reasoning would apply to `ERC20`, though in that case I guess we missed it due to it having a default constructor. WDYT?

nventuro commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+    {
+        address implementer = _erc1820.getInterfaceImplementer(to, TOKENS_RECIPIENT_INTERFACE_HASH);
+        if (implementer == address(0)) {
+            return(!to.isContract());
```
From my understanding of the following bit of the EIP, we should `require` the value that this function returns (which is done on `_mint` but not `_send`):

>The token contract MUST revert when sending in any of the following cases:
    * The recipient is a contract, and it does not implement the ERC777TokensRecipient interface via ERC1820.

nventuro commented:
Things are in a good place right now, but there are still some improvements to be done / missing pieces.

* We need `TokensSender` and `TokensRecipient` tests (this shouldn't be too hard)
* We need to go over the EIP text and see if there are any missing test cases (e.g. right now we're missing a `require` on the result of `_callTokensReceived`)
* We need to add tests to the internal functions (`_send`, `_burn` and `_mint`)
* We need to decide if we want to provide a simpler API (e.g. removing the `data` parameter via overloading). This can be postponed however.

tinchoabbate commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+
+    uint256 private _granularity;
+
+    bytes32 constant private TOKENS_SENDER_INTEFACE_HASH = keccak256("ERC777TokensSender");
```
Small typo: `TOKENS_SENDER_INTEFACE_HASH` should be `TOKENS_SENDER_INTERFACE_HASH`

tinchoabbate commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+    )
+    private
+    {
+        address implementer = _erc1820.getInterfaceImplementer(from, TOKENS_SENDER_INTEFACE_HASH);
```
Small typo: `TOKENS_SENDER_INTEFACE_HASH` should be `TOKENS_SENDER_INTERFACE_HASH`

nventuro commented:
Tests are failing because the `sender` for [this test](https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1684/files#diff-2a595e967394ea074e615e1f5fa2e3a0R254) is dynamic (i.e. it is the hook contract's address), so we need to put it inside an object so that it can be properly mutated on each `beforeEach` block. I'm not sure what's the best way to do this though, I'm not a fan of putting it inside `this`, or a `config` object that's passed to the helper, etc. @frangio ideas?

frangio commented:
I was able to get coverage running with an updated fork of `ganache-cli-coverage`. Please see https://github.com/frangio/openzeppelin-solidity/commit/42554ccb42fab8ac1026f18dfe06bd6d7e46a92f. I wasn't able to push the commit directly to this PR's branch. (What's up with that?)

---

@nventuro Oh, I see what the problem is. Is it necessary to deploy a new sender for each test? We can also just deploy one for the whole `context` in a `before` block.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+    * @dev Authorize an operator for the sender
+    * @param operator address to be authorized as operator
+     */
+    function authorizeOperator(address operator) public {
```
I don't think we've done this in other contracts yet, but I want us to consider keeping this function `external` only, and having a separate `internal` function for extending the contract etc.

@nventuro What do you think?

nventuro replied:
Yes - I wouldn't make the `internal` one just yet to keep this PR as simple as possible, but making these `external` sounds good.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+     */
+    function authorizeOperator(address operator) public {
+        require(msg.sender != operator);
+        if (_defaultOperators[operator]) {
```
I wouldn't have each branch in a separate private function. The only thing different is one statement. The event logged is exactly the same. Same for `revokeOperator`.

nventuro replied:
Good catch.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+        require((amount % _granularity) == 0);
+
+        // revert if 'to' is a contract not implementing tokensReceived()
+        require(_callTokensReceived(operator, address(0), to, amount, userData, operatorData));
```
The ordering of this is probably wrong. Although the EIP is not explicit about it for minting, it's very clear in `send` that the `tokensReceived` hook has to be called after updating the balance:

>The token contract MUST call the `tokensReceived` hook after updating the state.

nventuro replied:
It actually is explicit:

>>>The following rules apply when calling the tokensReceived hook:
    The tokensReceived hook MUST be called every time the balance is incremented.
    The tokensReceived hook MUST be called after the state is updated—i.e. after the balance is incremented.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+        address tokenHolder
+    ) public view returns (bool) {
+        return operator == tokenHolder ||
+            _defaultOperators[operator] && !_revokedDefaultOperators[tokenHolder][operator] ||
```
I'd add parentheses around the `x && y` expression for clarity.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+        // Update state variables
+        _totalSupply = _totalSupply.add(amount);
+        _balances[to] = _balances[to].add(amount);
+        assert((_balances[to] % _granularity) == 0);
```
I'm not a fan of these `assert`s.

nventuro replied:
Haha at this point I only add these to see if you'll eventually cave.

frangio commented on contracts/drafts/ERC777/ERC777.sol:
```diff
@@ -0,0 +1,402 @@
+        assert((_balances[from] % _granularity) == 0);
+        assert((_balances[to] % _granularity) == 0);
+
+        _callTokensReceived(operator, from, to, amount, userData, operatorData);
```
`_callTokensReceived` returns a boolean value that is being ignored. :upside_down_face:

We should remove that return value and make the function revert itself, just like `_callTokensToSend`.

nventuro commented:
All that's missing now are the `_mint` tests (which should be easy to add to the send/burn blocks).

frangio commented on test/drafts/ERC777/ERC777.test.js:
```diff
@@ -0,0 +1,448 @@
+    shouldInternalMintTokens(anyone, defaultOperatorA, granularity, data, operatorData);
+    shouldInternalMintTokens(anyone, defaultOperatorA, granularity.muln(2), data, operatorData);
+
+    it('reverts when sending an amount non-multiple of the granularity', async function () {
```
```suggestion
    it('reverts when minting an amount non-multiple of the granularity', async function () {
```

frangio commented:
For reference, I measured the gas cost of a `send` to an account (EOA) with no hooks registered.

- No token balance in the receiver: 60695 gas
- Some token balance in the receiver: 45695 gas
